langchain/agents/agent_toolkits/openapi/planner.py

```diff
@@ -0,0 +1,375 @@
+        # print(f"Request tool input: {data}\nResponse: {response}")
+        return self.llm_chain.predict(
+            response=response, instructions=data["output_instructions"]
+        ).strip()
+
+    async def _arun(self, text: str) -> str:
+        pass
+
+
+class RequestsPostToolWithSummary(BaseRequestsTool, BaseTool):
+    """Tool for making a POST request to an API endpoint."""
+
+    name = "requests_post"
+    description = """Use this when you want to POST to a website.
+Input to the tool should be a json string with 3 keys: "url", "data", and "output_instructions".
+The value of "url" should be a string.
+The value of "data" should be a dictionary of key-value pairs you want to POST to the url.
+The value of "summary_instructions" should be instructions on what information to extract from the response, for example the id(s) for a resource(s) that the POST request creates.
+Always use double quotes for strings in the json string."""
+
+    response_length: Optional[int] = MAX_RESPONSE_LENGTH
+    llm_chain = llm_chain = LLMChain(
+        llm=DEFAULT_LLM,
+        prompt=PromptTemplate(
+            template="""Here is an API response:\n\n{response}\n\n====
+Your task is to extract some information according to these instructions: {instructions}
+When working with API objects, you should usually use ids over names. Do not return any ids or names that are not in the response.
+If the response indicates an error, you should instead output a summary of the error.
+
+Output:""",
+            input_variables=["response", "instructions"],
+        ),
+    )
+
+    def _run(self, text: str) -> str:
+        try:
+            data = json.loads(text)
+            response = self.requests_wrapper.post(data["url"], data["data"])
+            response = response[: self.response_length]
+            # print(f"Request tool input: {data}\nResponse: {response}")
+            return self.llm_chain.predict(
+                response=response, instructions=data["output_instructions"]
+            ).strip()
+        except Exception as e:
+            return repr(e)
+
+    async def _arun(self, text: str) -> str:
+        pass
+
+
+#
+# Orchestrator, planner, controller.
+#
+def _create_api_planner_tool(
+    api_spec: ReducedOpenAPISpec, llm: BaseLanguageModel = DEFAULT_LLM
+) -> Tool:
+    endpoint_descriptions = [
+        f"{name} {description}" for name, description, _ in api_spec.endpoints
+    ]
+    prompt = PromptTemplate(
+        template=API_PLANNER_PROMPT,
+        input_variables=["query"],
+        partial_variables={"endpoints": "- " + "- ".join(endpoint_descriptions)},
+    )
+    chain = LLMChain(llm=llm, prompt=prompt)
+    tool = Tool(
+        name=API_PLANNER_TOOL_NAME,
+        description=API_PLANNER_TOOL_DESCRIPTION,
+        func=chain.run,
+    )
+    return tool
+
+
+def _create_api_controller_agent(
+    api_url: str,
+    api_docs: str,
+    requests_wrapper: RequestsWrapper,
+    llm: BaseLanguageModel = DEFAULT_LLM,
```

Comment:
Would it work if we didn't assign a default LLM in the creation of the controller / api planner and instead set the default (if we want a default) in the calling function? (so we can bubble the default LLM up through the call stack?

Reply:
done👍